process.go

```diff
@@ -123,91 +163,92 @@ type ProcessEventScope interface {
+	// The engine MAY allow re-beginning a process instance that
+	// has ended. Callers SHOULD assume that such behavior is unavailable.
 	End()
 
 	// Root returns the root of the targeted process instance.
 	//
-	// It panics if the instance has not been begun, or was begun but has
-	// subsequently been ended.
+	// It MUST NOT be called if the instance has not begun, or has ended.
 	Root() ProcessRoot
 
 	// ExecuteCommand executes a command as a result of the event or timeout
 	// message being handled.
 	//
-	// It panics if the instance has not been begun, or was begun but has
-	// subsequently been ended.
+	// It MUST NOT be called with a message of any type that has not been
+	// configured for production by a prior call to Configure().
+	//
+	// It MUST NOT be called if the instance has not begun, or has ended.
 	ExecuteCommand(m Message)
 
 	// ScheduleTimeout schedules a timeout message to be returned to the process
 	// at a specific time.
 	//
 	// Any pending timeout messages are cancelled when the instance is ended.
 	//
-	// It panics if the instance has not been begun, or was begun but has
-	// subsequently been ended.
+	// It MUST NOT be called if the instance has not begun, or has ended.
 	ScheduleTimeout(m Message, t time.Time)
 
-	// Log records an informational message within the context of the event being
-	// handled.
-	//
-	// The log message should be worded such that it makes sense to anyone familiar
-	// with the business domain.
+	// Log records an informational message within the context of the message
+	// that is being handled.
 	Log(f string, v ...interface{})
 }
 
 // ProcessTimeoutScope is an interface implemented by the engine and used by the
 // application to perform operations within the context of handling a timeout
 // message.
 type ProcessTimeoutScope interface {
-	// InstanceID is the ID of the targeted process instance.
+	// InstanceID returns the ID of the targeted process instance.
 	InstanceID() string
 
 	// End terminates the targeted process instance.
 	//
-	// After it has been called none of Root(), ExecuteCommand() or
-	// ScheduleTimeout() can be called within this scope or the scope of any future
-	// event or timeout that targets the same instance.
+	// After it has been called Root(), ExecuteCommand() and ScheduleTimeout()
+	// MUST NOT be called within this scope or the scope of any future message
+	// that targets the same instance.
 	//
-	// It panics if the target instance has not been begun.
+	// It MUST NOT be called if the instance has not begun.
 	//
-	// The precise semantics of ending a process instance are implementation
-	// defined. The engine is not required to allow re-beginning a process
-	// instance that has been ended.
+	// The engine MUST discard any timeout messages associated with this
+	// instance.
+	//
+	// The engine MAY allow re-beginning a process instance that
+	// has ended. Callers SHOULD assume that such behavior is unavailable.
```

Comment:
This "has" could have fit on the previous line? Could try reflow this doc block again.